cabal-install/Distribution/Client/IndexUtils/Timestamp.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Distribution.Client.IndexUtils.Timestamp
+-- Copyright   :  (c) Herbert Valerio Riedel
+-- License     :  BSD3
+--
+-- Timestamp type used in package indexes
+
+module Distribution.Client.IndexUtils.Timestamp
+    ( Timestamp
+    , nullTimestamp
+    , epochTimeToTimestamp
+    , timestampToUTCTime
+    , utcTimeToTimestamp
+    ) where
+
+import qualified Codec.Archive.Tar.Entry    as Tar
+import           Control.DeepSeq
+import           Data.Char                  (isDigit)
+import           Data.Int                   (Int64)
+import           Data.Time                  (UTCTime (..), fromGregorianValid,
+                                             makeTimeOfDayValid, showGregorian,
+                                             timeOfDayToTime, timeToTimeOfDay)
+import           Data.Time.Clock.POSIX      (posixSecondsToUTCTime,
+                                             utcTimeToPOSIXSeconds)
+import           Distribution.Compat.Binary
+import qualified Distribution.Compat.ReadP  as ReadP
+import           Distribution.Text
+import qualified Text.PrettyPrint           as Disp
+
+-- | UNIX timestamp (expressed in seconds since unix epoch, i.e. 1970).
+newtype Timestamp = TS Int64 -- Tar.EpochTime
+                  deriving (Eq,Ord,Enum,NFData)
+
+epochTimeToTimestamp :: Tar.EpochTime -> Timestamp
+epochTimeToTimestamp = TS
+
+timestampToUTCTime :: Timestamp -> UTCTime
+timestampToUTCTime (TS t) = posixSecondsToUTCTime (fromIntegral t)
+
+utcTimeToTimestamp :: UTCTime -> Maybe Timestamp
+utcTimeToTimestamp = Just . TS . round . utcTimeToPOSIXSeconds
+
+-- | Pretty-prints 'Timestamp' in ISO8601/RFC3339 format
+-- (e.g. @"2017-12-31T23:59:59Z"@)
+--
+-- Returns empty string for 'nullTimestamp'
+showTimestamp :: Timestamp -> String
+showTimestamp ts
+  | ts == nullTimestamp = ""
+  | otherwise = showGregorian utctDay ++ "T" ++ tod ++ "Z"
+  where
+    UTCTime{..} = timestampToUTCTime ts
+    tod = show . timeToTimeOfDay $ utctDayTime
+
+instance Binary Timestamp where
+    put (TS t) = put t
+    get = TS <$> get
+
+instance Text Timestamp where
+    disp = Disp.text . showTimestamp
+
+    parse = parsePosix ReadP.+++ parseUTC
+      where
+        -- | Parses unix timestamps, e.g. @"\@1474626019"@
+        parsePosix = do
+            _ <- ReadP.char '@'
+            t <- parse -- FIXME
+            pure (TS $ fromIntegral (t :: Int))
+
+        -- | Parses ISO8601/RFC3339-style UTC timestamps,
+        -- e.g. @"2017-12-31T23:59:59Z"@
+        --
+        -- TODO: support numeric tz offsets; allow to leave off seconds
+        parseUTC = do
+            ye <- parse -- TODO
```

Comment:
Note: The TODO here is about rejecting years with less than 4 digits as e.g. RFC3339 also discourages/obsoletes the use of 2-digit years since they are error-prone. I'll handle this when I extend the parsing to handle a few more input formats which can be normalised to UTC w/o requiring `IO` (see TODO two lines earlier).